aws_xray_sdk/core/serverless_context.py

```diff
@@ -0,0 +1,131 @@
+import os
+import logging
+
+from .models.facade_segment import FacadeSegment
+from .models.segment import Segment
+from .models.mimic_segment import MimicSegment
+from .context import CXT_MISSING_STRATEGY_KEY
+from .lambda_launcher import LambdaContext
+from .context import Context
+
+
+log = logging.getLogger(__name__)
+
+
+class ServerlessContext(LambdaContext):
+    """
+    Context used specifically for running middlewares on Lambda through the
+    Serverless design. This context is built on top of the LambdaContext, but
+    creates a Segment masked as a Subsegment known as a MimicSegment underneath
+    the Lambda-generated Facade Segment. This ensures that middleware->recorder's
+    consequent calls to "put_segment()" will not throw exceptions but instead create
+    subsegments underneath the lambda-generated segment. This context also
+    ensures that FacadeSegments exist through underlying calls to _refresh_context().
+    """
+    def __init__(self, context_missing='RUNTIME_ERROR'):
+        super(ServerlessContext, self).__init__()
+
+        strategy = os.getenv(CXT_MISSING_STRATEGY_KEY, context_missing)
+        self._context_missing = strategy
+
+    def put_segment(self, segment):
```

Comment:
My bold guess is that if you just use LambdaContext's put_subsegment on the "mimicSegment" everything will just work. To me the final json doc on a webapp+lambda is just one of many forms of what the SDKs are creating. The webapp layer trace entity is just THE giant local subsegment which is the single child of facade segment and single parent of many other subsegments. You can easily generate such a special form by just having a capture block on the entire lambda handler function and add some http metadata to that. The design can be simplified if you look at the problem we are trying to solve from a different view.

Reply:
PutSubsegment call to the Lambda Context would implicitly call parent_segment.increment(), and in this case, Facade Segments will be the parent, an FacadeSegmentMutationException would be thrown.